File: mobile_apps/ios_legacy/Runner/Services/hardware_wallet/HardwareWalletService.swift
//
//  HardwareWalletService.swift
//  TigerWallet
//
//  Complete Hardware Wallet Service for iOS
//

import Foundation

class HardwareWalletService {
    static let shared = HardwareWalletService()
    private let baseURL = "http://localhost:8443/api/v1/hardware"
    
    private init() {}
    
    // MARK: - Supported Devices
    
    func getSupportedDevices() -> [HardwareWalletDevice] {
        return [
            HardwareWalletDevice(id: "ledger", name: "Ledger", icon: "ledger_icon", supportedChains: ["ethereum", "bitcoin", "solana", "polygon", "bsc"]),
            HardwareWalletDevice(id: "trezor", name: "Trezor", icon: "trezor_icon", supportedChains: ["ethereum", "bitcoin"]),
            HardwareWalletDevice(id: "keepkey", name: "KeepKey", icon: "keepkey_icon", supportedChains: ["ethereum", "bitcoin"]),
            HardwareWalletDevice(id: "coldcard", name: "ColdCard", icon: "coldcard_icon", supportedChains: ["bitcoin"]),
            HardwareWalletDevice(id: "bitbox02", name: "BitBox02", icon: "bitbox02_icon", supportedChains: ["ethereum", "bitcoin"])
        ]
    }
    
    // MARK: - Connect
    
    func connect(deviceId: String, completion: @escaping (Result<HardwareWalletConnection, Error>) -> Void) {
        let url = URL(string: "\(baseURL)/connect")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body: [String: Any] = ["deviceId": deviceId]
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let data = data else {
                completion(.failure(NSError(domain: "HardwareWalletService", code: -1, userInfo: [NSLocalizedDescriptionKey: "No data"])))
                return
            }
            
            do {
                let connection = try JSONDecoder().decode(HardwareWalletConnection.self, from: data)
                completion(.success(connection))
            } catch {
                completion(.failure(error))
            }
        }.resume()
    }
    
    // MARK: - Get Address
    
    func getAddress(deviceId: String, chain: String, path: String, completion: @escaping (Result<String, Error>) -> Void) {
        let url = URL(string: "\(baseURL)/address")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body: [String: Any] = ["deviceId": deviceId, "chain": chain, "path": path]
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let address = json["address"] as? String else {
                completion(.failure(NSError(domain: "HardwareWalletService", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to get address"])))
                return
            }
            
            completion(.success(address))
        }.resume()
    }
    
    // MARK: - Sign Transaction
    
    func signTransaction(deviceId: String, txData: String, chain: String, completion: @escaping (Result<String, Error>) -> Void) {
        let url = URL(string: "\(baseURL)/sign")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body: [String: Any] = ["deviceId": deviceId, "txData": txData, "chain": chain]
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let signature = json["signature"] as? String else {
                completion(.failure(NSError(domain: "HardwareWalletService", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to sign"])))
                return
            }
            
            completion(.success(signature))
        }.resume()
    }
    
    // MARK: - Sign Message
    
    func signMessage(deviceId: String, message: String, completion: @escaping (Result<String, Error>) -> Void) {
        let url = URL(string: "\(baseURL)/sign-message")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body: [String: Any] = ["deviceId": deviceId, "message": message]
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let signature = json["signature"] as? String else {
                completion(.failure(NSError(domain: "HardwareWalletService", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to sign"])))
                return
            }
            
            completion(.success(signature))
        }.resume()
    }
    
    // MARK: - Disconnect
    
    func disconnect(deviceId: String, completion: @escaping (Result<Bool, Error>) -> Void) {
        let url = URL(string: "\(baseURL)/disconnect")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body: [String: Any] = ["deviceId": deviceId]
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        
        URLSession.shared.dataTask(with: request) { _, _, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            completion(.success(true))
        }.resume()
    }
}

struct HardwareWalletDevice {
    let id: String
    let name: String
    let icon: String
    let supportedChains: [String]
}

struct HardwareWalletConnection: Codable {
    let deviceId: String
    let address: String
    let publicKey: String
    let chain: String
    let connectedAt: Date
}

File: mobile_apps/android_legacy/app/src/main/java/com/tigerwallet/app/services/hardware_wallet/HardwareWalletService.java
package com.tigerwallet.app.services.hardware_wallet;

import android.content.Context;
import android.hardware.usb.UsbManager;
import org.json.JSONArray;
import org.json.JSONObject;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * Hardware Wallet Service for Android
 * Supports Ledger, Trezor, KeepKey, ColdCard, BitBox02
 */
public class HardwareWalletService {
    private static final String TAG = "HardwareWalletService";
    private static final String BASE_URL = "https://api.tigerwallet.com/v1/hardware";
    
    private final Context context;
    private final ExecutorService executor;
    private UsbManager usbManager;
    private Map<String, HardwareDevice> connectedDevices;
    
    public interface Callback<T> {
        void onSuccess(T result);
        void onError(Exception error);
    }
    
    public HardwareWalletService(Context context) {
        this.context = context;
        this.executor = Executors.newFixedThreadPool(4);
        this.connectedDevices = new HashMap<>();
        this.usbManager = (UsbManager) context.getSystemService(Context.USB_SERVICE);
    }
    
    /**
     * Get list of supported hardware wallet devices
     */
    public List<SupportedDevice> getSupportedDevices() {
        List<SupportedDevice> devices = new ArrayList<>();
        
        devices.add(new SupportedDevice("ledger", "Ledger", 
            new String[]{"ethereum", "bitcoin", "solana", "polygon", "bsc"}));
        devices.add(new SupportedDevice("trezor", "Trezor", 
            new String[]{"ethereum", "bitcoin"}));
        devices.add(new SupportedDevice("keepkey", "KeepKey", 
            new String[]{"ethereum", "bitcoin"}));
        devices.add(new SupportedDevice("coldcard", "ColdCard", 
            new String[]{"bitcoin"}));
        devices.add(new SupportedDevice("bitbox02", "BitBox02", 
            new String[]{"ethereum", "bitcoin"}));
        
        return devices;
    }
    
    /**
     * Connect to a hardware wallet device
     */
    public void connect(String deviceId, Callback<HardwareConnection> callback) {
        executor.execute(() -> {
            try {
                JSONObject body = new JSONObject();
                body.put("deviceId", deviceId);
                
                JSONObject response = makePostRequest("/connect", body);
                
                HardwareConnection connection = new HardwareConnection();
                connection.deviceId = response.getString("deviceId");
                connection.address = response.getString("address");
                connection.publicKey = response.getString("publicKey");
                connection.chain = response.getString("chain");
                connection.connectedAt = response.getLong("connectedAt");
                
                connectedDevices.put(connection.deviceId, new HardwareDevice(deviceId, connection.address));
                
                callback.onSuccess(connection);
            } catch (Exception e) {
                callback.onError(e);
            }
        });
    }
    
    /**
     * Get address from hardware wallet
     */
    public void getAddress(String deviceId, String chain, String path, Callback<String> callback) {
        executor.execute(() -> {
            try {
                JSONObject body = new JSONObject();
                body.put("deviceId", deviceId);
                body.put("chain", chain);
                body.put("path", path);
                
                JSONObject response = makePostRequest("/address", body);
                callback.onSuccess(response.getString("address"));
            } catch (Exception e) {
                callback.onError(e);
            }
        });
    }
    
    /**
     * Sign transaction with hardware wallet
     */
    public void signTransaction(String deviceId, String txData, String chain, Callback<String> callback) {
        executor.execute(() -> {
            try {
                JSONObject body = new JSONObject();
                body.put("deviceId", deviceId);
                body.put("txData", txData);
                body.put("chain", chain);
                
                JSONObject response = makePostRequest("/sign", body);
                callback.onSuccess(response.getString("signature"));
            } catch (Exception e) {
                callback.onError(e);
            }
        });
    }
    
    /**
     * Sign message with hardware wallet
     */
    public void signMessage(String deviceId, String message, Callback<String> callback) {
        executor.execute(() -> {
            try {
                JSONObject body = new JSONObject();
                body.put("deviceId", deviceId);
                body.put("message", message);
                
                JSONObject response = makePostRequest("/sign-message", body);
                callback.onSuccess(response.getString("signature"));
            } catch (Exception e) {
                callback.onError(e);
            }
        });
    }
    
    /**
     * SignTypedData for EIP-712
     */
    public void signTypedData(String deviceId, String domain, String message, Callback<String> callback) {
        executor.execute(() -> {
            try {
                JSONObject body = new JSONObject();
                body.put("deviceId", deviceId);
                body.put("domain", domain);
                body.put("message", message);
                
                JSONObject response = makePostRequest("/sign-typed-data", body);
                callback.onSuccess(response.getString("signature"));
            } catch (Exception e) {
                callback.onError(e);
            }
        });
    }
    
    /**
     * Get device firmware version
     */
    public void getFirmwareVersion(String deviceId, Callback<FirmwareInfo> callback) {
        executor.execute(() -> {
            try {
                JSONObject response = makeGetRequest("/firmware/" + deviceId);
                
                FirmwareInfo info = new FirmwareInfo();
                info.version = response.getString("version");
                info.bootloaderVersion = response.getString("bootloaderVersion");
                info.isUpdateAvailable = response.getBoolean("isUpdateAvailable");
                
                callback.onSuccess(info);
            } catch (Exception e) {
                callback.onError(e);
            }
        });
    }
    
    /**
     * Disconnect from hardware wallet
     */
    public void disconnect(String deviceId, Callback<Boolean> callback) {
        executor.execute(() -> {
            try {
                JSONObject body = new JSONObject();
                body.put("deviceId", deviceId);
                
                makePostRequest("/disconnect", body);
                connectedDevices.remove(deviceId);
                
                callback.onSuccess(true);
            } catch (Exception e) {
                callback.onError(e);
            }
        });
    }
    
    /**
     * Get connected devices
     */
    public Map<String, HardwareDevice> getConnectedDevices() {
        return new HashMap<>(connectedDevices);
    }
    
    /**
     * Verify address on device display
     */
    public void verifyAddress(String deviceId, String address, Callback<Boolean> callback) {
        executor.execute(() -> {
            try {
                JSONObject body = new JSONObject();
                body.put("deviceId", deviceId);
                body.put("address", address);
                
                JSONObject response = makePostRequest("/verify-address", body);
                callback.onSuccess(response.getBoolean("verified"));
            } catch (Exception e) {
                callback.onError(e);
            }
        });
    }
    
    private JSONObject makePostRequest(String endpoint, JSONObject body) throws Exception {
        URL url = new URL(BASE_URL + endpoint);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("POST");
        conn.setRequestProperty("Content-Type", "application/json");
        conn.setDoOutput(true);
        
        conn.getOutputStream().write(body.toString().getBytes(StandardCharsets.UTF_8));
        
        BufferedReader reader = new BufferedReader(
            new InputStreamReader(conn.getInputStream())
        );
        
        StringBuilder response = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) {
            response.append(line);
        }
        reader.close();
        
        return new JSONObject(response.toString());
    }
    
    private JSONObject makeGetRequest(String endpoint) throws Exception {
        URL url = new URL(BASE_URL + endpoint);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("GET");
        
        BufferedReader reader = new BufferedReader(
            new InputStreamReader(conn.getInputStream())
        );
        
        StringBuilder response = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) {
            response.append(line);
        }
        reader.close();
        
        return new JSONObject(response.toString());
    }
    
    // Data classes
    public static class SupportedDevice {
        public String id;
        public String name;
        public String[] supportedChains;
        
        public SupportedDevice(String id, String name, String[] chains) {
            this.id = id;
            this.name = name;
            this.supportedChains = chains;
        }
    }
    
    public static class HardwareDevice {
        public String deviceId;
        public String address;
        
        public HardwareDevice(String deviceId, String address) {
            this.deviceId = deviceId;
            this.address = address;
        }
    }
    
    public static class HardwareConnection {
        public String deviceId;
        public String address;
        public String publicKey;
        public String chain;
        public long connectedAt;
    }
    
    public static class FirmwareInfo {
        public String version;
        public String bootloaderVersion;
        public boolean isUpdateAvailable;
    }
}
